File: GLLara/GLLModelBone.m
//
//  GLLModelBone.m
//  GLLara
//
//  Created by Torsten Kammer on 01.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLModelBone.h"

#import "GLLASCIIScanner.h"
#import "GLLModel.h"
#import "simd_matrix.h"
#import "TRInDataStream.h"
#import "TROutDataStream.h"

@implementation GLLModelBone

- (instancetype)initWithModel:(GLLModel *)model;
{
    if (!(self = [super init])) return nil;
    
    _model = model;
    
    _parentIndex = UINT16_MAX;
    _children = [NSMutableArray array];
    
    _positionX = 0;
    _positionY = 0;
    _positionZ = 0;
    
    _positionMatrix = matrix_identity_float4x4;
    _inversePositionMatrix = matrix_identity_float4x4;
    
    _name = NSLocalizedString(@"Root bone", @"Only bone in a boneless format");
    
    return self;
    
}

- (instancetype)initFromSequentialData:(id)stream partOfModel:(GLLModel *)model atIndex:(NSUInteger)index error:(NSError *__autoreleasing*)error;
{
    if (!(self = [super init])) return nil;
    
    if (![stream isValid])
    {
        if (error)
            *error = [NSError errorWithDomain:GLLModelLoadingErrorDomain code:GLLModelLoadingError_PrematureEndOfFile userInfo:@{
                                                                                                                                 NSLocalizedDescriptionKey : NSLocalizedString(@"The file is missing some data.", @"Premature end of file error"),
                                                                                                                                 NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString(@"The file breaks off in the middle of the bones section. Maybe it is damaged?", @"Premature end of file error")
                                                                                                                                 }];
        return nil;
    }
    
    _model = model;
    
    _name = [stream readPascalString];
    _parentIndex = [stream readUint16];
    _positionX = [stream readFloat32];
    _positionY = [stream readFloat32];
    _positionZ = [stream readFloat32];
    
    _children = [NSMutableArray array];
    
    if (_parentIndex == index) {
        if ([_name hasPrefix:@"unused"]) {
            // Apparently that's a thing that people do. Create unused bones with themselves set as parent. Why, though? What's wrong with them?
            NSLog(@"Bone %lu (named \"%@\") has itself as parent. Unused, so set as root bone. Why do people do that?", index, _name);
            _parentIndex = UINT16_MAX;
        } else {
            if (error)
                *error = [NSError errorWithDomain:GLLModelLoadingErrorDomain code:GLLModelLoadingError_CircularReference userInfo:@{
                                                                                                                                    NSLocalizedDescriptionKey : [NSString stringWithFormat:NSLocalizedString(@"Bone \"%@\" has itself as an ancestor.", @"Found a circle in the bone relationships."), self.name],
                                                                                                                                    NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString(@"The bones would form an infinite loop.", @"Found a circle in a bone relationship")}];
            return nil;
        }
    }
    
    if (![stream isValid])
    {
        if (error)
            *error = [NSError errorWithDomain:GLLModelLoadingErrorDomain code:GLLModelLoadingError_PrematureEndOfFile userInfo:@{
                                                                                                                                 NSLocalizedDescriptionKey : NSLocalizedString(@"The file is missing some data.", @"Premature end of file error"),
                                                                                                                                 NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString(@"The file breaks off in the middle of the bones section. Maybe it is damaged?", @"Premature end of file error")
                                                                                                                                 }];
        return nil;
    }
    
    _positionMatrix = simd_mat_positional(simd_make_float4(_positionX, _positionY, _positionZ, 1.0f));
    _inversePositionMatrix = simd_mat_positional(simd_make_float4(-_positionX, -_positionY, -_positionZ, 1.0f));
    
    return self;
}

- (GLLModelBone *)parent
{
    if (self.parentIndex >= self.model.bones.count) return nil;
    return self.model.bones[self.parentIndex];
}

#pragma mark - Export

- (NSString *)writeASCII;
{
    NSMutableString *result = [NSMutableString string];
    
    [result appendFormat:@"%@\n", self.name];
    [result appendFormat:@"%d\n", self.parentIndex != NSNotFound ? (int) self.parentIndex : -1];
    [result appendFormat:@"%f %f %f\n", self.positionX, self.positionY, self.positionZ];
    
    return [result copy];
}

- (NSData *)writeBinary;
{
    TROutDataStream *stream = [[TROutDataStream alloc] init];
    
    [stream appendPascalString:self.name];
    [stream appendUint16:(uint16_t) self.parentIndex];
    [stream appendFloat32:self.positionX];
    [stream appendFloat32:self.positionY];
    [stream appendFloat32:self.positionZ];
    
    return stream.data;
}

@end

File: GLLara/GLLModelBone.swift
//
//  GLLModelBone.swift
//  GLLara
//
//  Created by Torsten Kammer on 19.02.23.
//  Copyright © 2023 Torsten Kammer. All rights reserved.
//

import Foundation

/*!
 * @abstract Description of a bone in a model.
 * @discussion A bone is a transformable entity; vertices belong to one or several bones, with different weights. The bone here is purely a static description and with default values. It does not contain any transformation information.
 */
@objc class GLLModelBone: NSObject {
    @objc let name: String
    @objc let parentIndex: Int
    @objc var position: simd_float3
    
    // Transformations for the bone
    @objc let inversePositionMatrix: matrix_float4x4
    @objc let positionMatrix: matrix_float4x4
    
    @objc var parent: GLLModelBone? = nil
    @objc var children: [GLLModelBone] = []
    
    override init() {
        name = NSLocalizedString("Root bone", comment: "Only bone in a boneless format")
        position = SIMD3(repeating: 0.0)
        parentIndex = -1
        positionMatrix = matrix_identity_float4x4
        inversePositionMatrix = matrix_identity_float4x4
        
        super.init()
    }
    
    init(sequentialData stream: GLLDataReader) throws {
        guard stream.isValid else {
            throw NSError(domain: GLLModelLoadingErrorDomain, code: Int(GLLModelLoadingErrorCode.prematureEndOfFile.rawValue), userInfo: [
                NSLocalizedDescriptionKey : NSLocalizedString("The file is missing some data.", comment: "Premature end of file error"),
                NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString("The file breaks off in the middle of the bones section. Maybe it is damaged?", comment: "Premature end of file error.")
            ])
        }
        
        name = stream.readPascalString()
        let parentIndexValue = stream.readInt16()
        parentIndex = (parentIndexValue == Int16.max) ? -1 : Int(parentIndexValue)
        let x = stream.readFloat32()
        let y = stream.readFloat32()
        let z = stream.readFloat32()
        position = SIMD3(x, y, z)
        
        positionMatrix = simd_mat_positional(SIMD4(position, 1.0))
        inversePositionMatrix = simd_mat_positional(SIMD4(-position, 1.0))
        
        guard stream.isValid else {
            throw NSError(domain: GLLModelLoadingErrorDomain, code: Int(GLLModelLoadingErrorCode.prematureEndOfFile.rawValue), userInfo: [
                NSLocalizedDescriptionKey : NSLocalizedString("The file is missing some data.", comment: "Premature end of file error"),
                NSLocalizedRecoverySuggestionErrorKey : NSLocalizedString("The file breaks off in the middle of the bones section. Maybe it is damaged?", comment: "Premature end of file error.")
            ])
        }
    }
    
    // Export
    func writeASCII() -> String {
        var result = ""
        
        result.append("\(name)\n")
        result.append("\(parentIndex)\n")
        result.append("\(position.x) \(position.y) \(position.z)\n")
        
        return result
    }
    
    func writeBinary() -> Data {
        let stream = TROutDataStream()
        
        stream.appendPascalString(name)
        stream.appendUint16(UInt16(truncating: parentIndex as NSNumber))
        stream.appendFloat32(position.x)
        stream.appendFloat32(position.y)
        stream.appendFloat32(position.z)
        
        return stream.data()
    }
}
